### ISSUE TITLE
Possibly incorrect stack effects documented for bi*, bi@, etc.

### ISSUE THREAD

**Issue by timor** [factor/factor#2613]
I think that technically, the stack effect for `bi*`, which is returned by `\ bi* stack-effect` as `( x y p q -- )` may be incorrect.

For example, consider `41 [ ] [ 1 + ] bi*`.  It works correctly on the empty stack, even though only three inputs have been provided.  I think the correct effect should be

`: bi* ( ..a y p: ( ..a -- ..b ) q: ( ..b y -- ..c ) -- ..c )`.

The description in the documentation basically describes it as equivalent to

`: bi* ( ..a x y p: ( ..a x -- ..b ) q: ( ..b y -- ..c ) -- )`,

which is a special case of the first one.

This would also extend to words like `bi@`. For example `42 [ 33 ] bi@` is valid code, even though only 2 instead of the 3 inputs of the defined stack effect are present.  However, in the case of `bi@`, it is not possible to completely express the correct effect in terms of row vars of the quotation effect, since it would have to look something like this:

`: bi@ ( ..a y quot: ( ..a -- ..b )/( ..b y -- ..c ) -- ..c )`.

So I guess technically the correct effect would be

`: bi@ ( ..a y quot -- ..c )`,

and the documentation could read something like "Applies the quotation to `..a`, producing `..b`, then to `..b y`, producing `..c`"?

**Comment by mrjbq7** [comment]
Makes sense.

This could fit into some improvements on our stack checker, but I hesitate to change the current definitions -- wouldn't it make them harder to understand for people newer to Factor?  But I also like the idea of making them more correct... Hmm..

**Comment by NoahTheDuke** [comment]
As a relative newbie to Factor, the current stack effects are much easier to understand.

Is there a reason the implementation doesn't conform to the current docs?

**Comment by mrjbq7** [comment]
This is one of the areas of the stack-checker that @erg and I have talked about improving in the development cycle after 0.99 is released, or if anyone gets interested in working on that sooner.

**Comment by mrjbq7** [comment]
Want to make a PR with fixed stack effects?

We could possibly have them commented out next to the current ones, or replacing them with correct ones if we all think it’s a good idea.

**Comment by timor** [comment]
I agree with @NoahTheDuke that the current ones are much easier to understand, since they match the "intended" meaning of the word better.

I guess these combinators basically define a whole family of sensible effects...

**Comment by nomennescio** [comment]
I always had the feeling when looking at stack effects that some shortcuts were taken, so I didn't find them that intuitive at all for the generic usecase, although I got used to them for common usecases.

Maybe we can solve this with an in-depth technical article on the actual correct stack effect declarations, and then introduce a "convenience" notation there, and keep all other articles according to that "convenience" notation (if possible of course, adapt where necessary).

**Comment by NoahTheDuke** [comment]
1) Such an article would be a boon in general. There's aren't many in depth articles about Factor and sobering like this would go a long way to help newcomers develop a sense of how things actually work.
2) Linking that article in all of the relevant pages would be great too.

To maybe say the obvious, documentation serves two purposes: tutorial/guide and reference. Racket has lots of good combinations of this, like the tutorial pages on macros that link into the reference/technical documentation, and then linking back from those reference docs to the tutorials/guides as well.

The fake-but-friendly stack effects are standing in for tutorial here when maybe they shouldn't.

**Comment by timor** [comment]
There is an article about stack effects in the manual, which I remember reading early on when getting to know Factor: https://docs.factorcode.org/content/article-effects.html

I think it pretty much covers all important details?

**Comment by mrjbq7** [comment]
I was thinking it would be good to have a PR that, for example, changes this:

```factor
: bi* ( x y p q -- )
    [ dip ] dip call ; inline
```

maybe to this with a comment:

```factor
: bi* ( x y p q -- ) ! ( ..a y p: ( ..a -- ..b ) q: ( ..b y -- ..c ) -- ..c )
    [ dip ] dip call ; inline
```

then we can at least track the more correct stack effect
